File: user_interface/src/contracts/workflows.ts
import type { ConfigurationProfile } from './configuration';

/** Immutable inputs for a retry-safe background operation. */
export interface WorkflowRequest {
  client_request_identifier: string;
  operation: 'train' | 'evaluate' | 'export' | 'prepare_dataset';
  experiment_identifier?: string;
  dataset_identifier?: string;
  checkpoint_identifier?: string;
  cpu_threads?: number;
  stop_after_step?: number;
  source_identifier?: string;
  source_subdirectory?: string;
  dataset_name?: string;
  export_name?: string;
  lightweight?: boolean;
}
export interface TrainingPreflight {
  allowed: boolean;
  blockers: string[];
  warnings: string[];
  resolved_settings: Record<string, unknown>;
  remaining_budget_seconds: number | null;
  remaining_experiment_slots: number | null;
}
export interface JobSnapshot {
  job_identifier: string;
  status:
    | 'queued'
    | 'running'
    | 'paused'
    | 'stopped'
    | 'completed'
    | 'cancelled'
    | 'failed'
    | 'interrupted'
    | 'needs_input';
  phase: string;
  configuration: ConfigurationProfile;
  available_actions: string[];
  created_at: string;
  updated_at: string;
  progress: number | null;
  estimated_seconds_remaining: number | null;
  operation?: string | null;
  experiment_identifier?: string | null;
  frozen_settings: Record<string, unknown>;
  result: Record<string, unknown> | null;
  error: Record<string, string> | null;
  metrics: Record<string, number>;
  requested_action: string | null;
  latest_event_identifier: number | null;
}
export interface JobList {
  items: JobSnapshot[];
}
export interface JobActionRequest {
  client_request_identifier: string;
  action: 'stop' | 'cancel' | 'pause' | 'resume';
}

export interface JobEvent {
  event_identifier: number;
  job_identifier: string;
  snapshot: JobSnapshot | null;
}

File: schemas/workflows.py
"""Browser workflow requests using registered references instead of paths."""

from typing import Literal

from pydantic import Field

from schemas.base import StrictRecord
from schemas.configuration import RequestIdentifier

WorkflowOperation = Literal["train", "evaluate", "export", "prepare_dataset"]


class WorkflowRequest(StrictRecord):
    """Resolve a bounded local operation on the server's original workspace."""

    client_request_identifier: RequestIdentifier
    operation: WorkflowOperation
    experiment_identifier: str = Field(
        default="cpu_experiment", pattern=r"^[a-z][a-z0-9_]{0,63}$"
    )
    dataset_identifier: str | None = Field(default=None, max_length=128)
    checkpoint_identifier: str | None = Field(default=None, max_length=128)
    cpu_threads: int = Field(default=4, ge=1, le=16)
    stop_after_step: int = Field(default=1000, ge=1, le=1000)
    source_identifier: str | None = Field(default=None, max_length=128)
    source_subdirectory: str = Field(default="", max_length=4096)
    dataset_name: str = Field(default="local_images", pattern=r"^[a-z][a-z0-9_]{0,63}$")
    export_name: str = Field(
        default="001V", pattern=r"^[A-Za-z0-9][A-Za-z0-9_.-]{0,63}$"
    )
    lightweight: bool = False


class WorkflowActionRequest(StrictRecord):
    """Make retries of supported actions safe across browser reloads."""

    client_request_identifier: RequestIdentifier
    action: Literal["stop", "cancel", "pause", "resume"]


class TrainingPreflight(StrictRecord):
    """Explain eligibility without creating files or reserving experiment time."""

    allowed: bool
    blockers: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    resolved_settings: dict[str, object] = Field(default_factory=dict)
    remaining_budget_seconds: float = Field(default=0, ge=0)
    remaining_experiment_slots: int = Field(default=0, ge=0)
